external/fv3fit/tests/training/test_main.py

```diff
@@ -302,6 +302,36 @@ def get_config(
     )
 
 
+def test_train_config_override_args(tmpdir, mock_load_batches, mock_train_dense_model):
+    model_type = "DenseModel"
+    hyperparameter_dict = {
+        "dense_network": {"width": 6},
+        "optimizer_config": {"name": "MyOpt", "kwargs": {"lr": 0.01}},
+    }
+    config = get_config(
+        tmpdir,
+        derived_output_variables=[],
+        model_type=model_type,
+        hyperparameter_dict=hyperparameter_dict,
+        use_validation_data=True,
+    )
+    mock_load_batches.return_value = [config.mock_dataset for _ in range(6)]
+    assert config.hyperparameters.dense_network["width"] == 6
+    assert config.hyperparameters.optimizer_config["kwargs"]["lr"] == 0.01
+    patch_args = [
+        "--hyperparameters.dense_network.width",
+        "12",
+        "--hyperparameters.optimizer_config.kwargs.lr",
+        "0.02",
+    ]
+    fv3fit.train.main(config.args, unknown_args=patch_args)
```

Comment:
Having to do the override argument testing and the `main` testing all together makes test feel too complex.  I think this would be the advantage of having some type of function which generally updates a config.  That way we wouldn't have to write this test every time?  Nothing to do now, but food for thought on something to work towards.

Reply:
I could split main into "get configs" and "call main which has configs as call signature", and mock out the latter function when testing the first. This is fairly close to what the test does right now - all training is mocked out.

You do have the "function which generally updates a config" already, and it's (presumably?) already tested, so arguably this test could be deleted altogether. But I want to test that `main` actually makes proper use of that function.

Reply:
Another way of saying some of this is, the "main" tests are testing that command-line arguments are properly propagated to code behavior. They aren't testing anything else (training, data loading, etc.) as it's all been mocked.

Reply:
>You do have the "function which generally updates a config" already, and it's (presumably?) already tested

Yep, you're right, the args -> dict is tested and general function.  

I guess the complexity here is driven by parsing args in the `__main__`, passing them into `main()` and then having to derive a `TrainConfig` from there.  I thought it was doing a larger integration test and also checking the config (missed that training is also mocked), whereas from the test name, I was just expecting something looking for a valid and updated `TrainConfig` based on arguments.  It's all coupled, so the test has to look like this.

Reply:
> Another way of saying some of this is, the "main" tests are testing that command-line arguments are properly propagated to code behavior.

Thanks for clarifying this.  Makes sense to me now.